### DISCUSSION TITLE
Pulumi and local development with Kubernetes

### DISCUSSION THREAD

**Discussion by vorotech** [pulumi/pulumi#7307] [General]
Hello and thank you for a great product!

I'd like to share the feedback on how do we do the local development and which role is taken by Pulumi. And ask you a few questions about whether we are doing it right or other conceptual ways to achieve the same things are designed.

The target environment is a Kubernetes cluster running in Docker Desktop (Windows/Mac) for developers and GKE cluster in Google Cloud in staging/production.
The Kubernetes cluster deployment is done by Pulumi only for staging/production, and it is managed as a separate stack, so any other services targeting the k8s has the clusterRef in their config.

```yaml
# Pulumi.staging.yaml
config:
  gcp:project: my-project
  kubernetes:clusterStack: my-org/my-cluster/staging
```
Take a look at the dev config we utilize to target the local k8s cluster instead:

```yaml
# Pulumi.dev.yaml
config:
  docker:skipPush: "true"
  kubernetes:isLocal: "true"
  kubernetes:context: docker-desktop
```
As you can see there few custom config variables that we need in order to change the behaviour of the following Pulumi script.

```ts
// Get k8s provider ref
const k8sConfig = new pulumi.Config('kubernetes');
let provider;
if (k8sConfig.getBoolean('isLocal')) {
  provider = new k8s.Provider('k8s', { context: k8sConfig.require('context') });
} else {
  const clusterStack = new pulumi.StackReference(k8sConfig.require('clusterStack'));
  provider = new k8s.Provider("k8s", { kubeconfig: clusterStack.getOutput("kubeConfig") });
}
```

Another statement is applied to the Docker image build process; because we want to avoid pushing it to remote GCR:

```ts
const skipPush = new pulumi.Config('docker').getBoolean('skipPush');
const image = new docker.Image(appName, {
  imageName: skipPush ? imageName : `${registryHostName}/${gcp.config.project}/${imageName}`,
  build: {
    context: '..'  // Build Dockerfile 
  },
  skipPush: skipPush
})
```

And the last, but not the least to use this image properly while creating the Pod we also should consider either it local or remote (`image.imageName` is the unique hash which was not pushed to local Docker):

```ts
// Defined a Pod
const pb = new kx.PodBuilder({
  containers: [
    {
      name: appName,
      image: skipPush ? imageName : image.imageName, // local vs remote image
      imagePullPolicy: 'IfNotPresent', // <-- important for local
      ports: { http: 80 },
    }
  ]
});
```

Another concern we have is when targeting local environment in this way, the ability to use remote persistence at the Pulumi service makes not much sense since it is valid only per single machine.

Looking forward to hearing the thoughts of community and Pulumi dev team.

**Comment by EvanBoyle** [comment]
@vorotech and anyone else interested in this issue:

I'm actually working on a project in this space and would be interested to show you a demo and collect some feedback and understand how you're using pulumi today. If you're interested, feel free to schedule a time that works for you: https://calendly.com/pulumi-labs/60-minutes

**Reply by nesl247** [reply]
I too would be interested. We're looking into having minikube for local dev, but I want to make use of the existing IaC if possible.

**Reply by sheldonhull** [reply]
@EvanBoyle any update, blog post, or repo example? I'm about to dive into some K8 deployment work and trying to figure out (I'm newer to K8) how to configure the workflow so that I could spin up local resources in minikube to test, and then flip to the remote using Pulumi. 

The local in particular is of interest, as I'd assume that the state management for a local dev environment is still in the Pulumi store, and tracked, when in reality I want this as a local ephemeral environment, and not something requiring remote state storage visible to all the team. Once I move to staging shared environment, then I'd expect to see this stored. 

If you have any links to docs on this topic that I've missed, I'd welcome that as well in case my understanding of the local/remote workflow is wrong. 
My best guess right now is using [filesystem](https://www.pulumi.com/docs/intro/concepts/state/) state storage override for local, and then staging defaults to Pulumi state storage.

**Comment by spara** [comment]
This would fantastic for this blog https://github.com/pulumi/blog-content/issues/2

**Comment by Sleepful** [comment]
to update the Docker image with a new tag I had to do this:


```makefile
#makefile
tag:
	pulumi config set kubernetes:multifrontTag `gdate +%F.%H.%M.%S`
```
(`gdate` or `date` depending on your OS)
```ts
export const k8sConfig = new pulumi.Config('kubernetes');
export const multifrontTag = k8sConfig.get('multifrontTag')

const imgName = `${multifrontName}-img:${multifrontTag}`
export const multifrontImg = new docker.Image(imgName, {
  imageName: skipPush ? imgName : interpolate`${reg.endpoint}/${imgName}`,
  build: {
    context: multifrontPath
  },
  skipPush: skipPush
})
```

**Reply by Sleepful** [reply]
then just run `make tag` whenever you want a new docker img deployed
